package-testing/android-relay/app/src/main/java/cloud/eppo/android/sdkrelay/TestClientActivity.kt

```diff
@@ -0,0 +1,277 @@
+        AssignmentRequest(
+            flag = requestObj.getString("flag"),
+            subjectKey = requestObj.getString("subjectKey"),
+            assignmentType = requestObj.getString("assignmentType"),
+            subjectAttributes = requestObj.getJSONObject("subjectAttributes"),
+            defaultValue = requestObj.get("defaultValue"))
+
+    // Ack function for responding to the server.
+    val ack = args.last() as Ack
+
+    val client = EppoClient.getInstance()
+
+    val eppoValues: MutableMap<String, EppoValue> = mutableMapOf()
+    assignmentRequest.subjectAttributes.keys().forEach {
+      val value = assignmentRequest.subjectAttributes.get(it.toString())
+      val typedValue: EppoValue =
+          when (value) {
+            is Int -> EppoValue.valueOf(value.toDouble())
+            is String -> EppoValue.valueOf(value)
+            is Double -> EppoValue.valueOf(value)
+            is Boolean -> EppoValue.valueOf(value)
+            is List<*> -> EppoValue.valueOf(value.map { insideIt -> insideIt.toString() })
+            else -> {
+              Log.e(TAG, "Invalid or null subject attribute type.")
+              EppoValue.nullValue()
+            }
+          }
+      eppoValues[it.toString()] = typedValue
+    }
+    val subjectAttributes = Attributes(eppoValues)
+
+    val result = getResult(assignmentRequest, client, subjectAttributes)
+    val jsonResult = JSONObject()
+    when (result) {
+      is JsonNode -> jsonResult.put("result", JSONObject(result.toString()))
+      else -> jsonResult.put("result", result)
+    }
+    ack.call(jsonResult.toString())
+  }
+
+  private fun handleDisconnect(args: Array<Any>) {
+    Log.d(TAG, "Disconnected")
+    status.postValue(getString(R.string.status_disconnected))
+
+    // Shut down the connection and remove listeners
+    SocketHandler.mSocket.disconnect()
+    SocketHandler.mSocket.off()
+
+    runOnUiThread { Toast.makeText(this, "Test runner disconnected", Toast.LENGTH_LONG).show() }
+  }
+
+  private fun handleReset(args: Array<Any>) {
+    Log.d(TAG, "SDK Reset")
+
+    // Initialize the SDK
+    val ack = args.last() as Ack
+    reInitializeEppoClient().thenAccept { ack.call(true) }
+  }
+
+  override fun onDestroy() {
+    super.onDestroy()
+
+    SocketHandler.mSocket.disconnect()
+    SocketHandler.mSocket.off()
+  }
+
+  private fun sendReady() {
+    SocketHandler.mSocket.emit("READY", arrayOf<String>(READY_PACKET)) {
+      Log.d(TAG, "Ready message acked")
+    }
+  }
+
+  private fun reInitializeEppoClient(): CompletableFuture<EppoClient> {
+    // Most of the settings used here are intended for debugging only.
+    return EppoClient.Builder(API_KEY, application)
+        .forceReinitialize(true)
+        .ignoreCachedConfiguration(true)
+        .host(EPPO_API_ADDRESS)
+        .isGracefulMode(false) // Debug: surface exceptions
+        .assignmentLogger(::assignmentLogger)
+        .buildAndInitAsync()
+  }
+
+  private fun getResult(
```

Comment:
more like `parseResult`

Reply:
The only parsing in this function is to wrestle the JSON defaultValue into the same json object type as is required by the client API.
maybe `getAssignmentFromClient` is more fitting?